rust-lang/rust
# Only report Self: Sized violation once.

## Background

The cause of #20692 seems to be in [middle::trait::object_safety::object_safety_violations](https://github.com/rust-lang/rust/blob/7ae4a8e9f3150f62964151ef54b3da3cd24ee123/src/librustc/middle/traits/object_safety.rs#L78-L85). The `supertrait_def_ids` call returns an iterator that iterates over trait definition ids. The iteration seems to start at the current trait and then traverse supertraits in order. For example, in the following block of code, it goes `Bar` -> `Foo` -> `Sized`:

``` rust
trait Foo: Sized { }
trait Bar: Foo { }

fn foo<T: Bar>(x: &T) {
    let _ = x as &Bar; //~ ERROR cannot convert to a trait object because trait `Bar` is not object-safe
}

fn main () {}
```

However for each trait in the hierarchy, `object_safety_violations_in_trait` gets called (in the `.flat_map()` call). Each call (all 3 of them) results in a `Self : Sized` violation since each trait in the hierarchy has `Sized` as a supertrait. That is why the above code compiles with 2 extra notes today:

```
: rustc 20692.rs
20692.rs:5:13: 5:14 error: cannot convert to a trait object because trait `Bar` is not object-safe [E0038]
20692.rs:5     let _ = x as &Bar; //~ ERROR cannot convert to a trait object because trait `Bar` is not object-safe
                       ^
20692.rs:5:13: 5:14 note: the trait cannot require that `Self : Sized`
20692.rs:5     let _ = x as &Bar; //~ ERROR cannot convert to a trait object because trait `Bar` is not object-safe
                       ^
20692.rs:5:13: 5:14 note: the trait cannot require that `Self : Sized`
20692.rs:5     let _ = x as &Bar; //~ ERROR cannot convert to a trait object because trait `Bar` is not object-safe
                       ^
20692.rs:5:13: 5:14 note: the trait cannot require that `Self : Sized`
20692.rs:5     let _ = x as &Bar; //~ ERROR cannot convert to a trait object because trait `Bar` is not object-safe
                       ^
```
## The fix

The fix I've submitted is to move the `trait_has_sized_self` check out of `object_safety_violations_in_trait` to `object_safety_violations` so it's only called once, and not for each supertrait.
## How to test?

I need help here. I don't see how to test it. This doesn't work:

``` rust
trait Foo: Sized { }
trait Bar: Foo { }

fn foo<T: Bar>(x: &T) {
    let _ = x as &Bar; //~ ERROR cannot convert to a trait object because trait `Foo` is not object-safe
                       //~^ NOTE the trait cannot require that `Self : Sized`
}

fn main() {}
```

Putting this in compile-fail passes without the fix. It seems that this doesn't try to match the number of notes exactly, so even though compilation results in 3 notes emitted today, it matches the one that is expected in the test and ignores the extra error output.

Any advice on how to test this fix?

Fixes #20692.
